rest/route/patch.go

```diff
@@ -127,6 +127,42 @@ func (p *patchByIdHandler) Run(ctx context.Context) gimlet.Responder {
 	return gimlet.NewJSONResponse(foundPatch)
 }
 
+////////////////////////////////////////////////////////////////////////
+//
+// GET /rest/v2/patches/{patch_id}/raw
+
+type patchRawHandler struct {
+	patchID    string
+	moduleName string
+	sc         data.Connector
+}
+
+func makePatchRawHandler(sc data.Connector) gimlet.RouteHandler {
+	return &patchByIdHandler{
+		sc: sc,
+	}
+}
+
+func (p *patchRawHandler) Factory() gimlet.RouteHandler {
+	return &patchByIdHandler{sc: p.sc}
+}
+
+func (p *patchRawHandler) Parse(ctx context.Context, r *http.Request) error {
+	p.patchID = gimlet.GetVars(r)["patch_id"]
+	p.moduleName = r.URL.Query().Get("module")
+
+	return nil
+}
+
+func (p *patchRawHandler) Run(ctx context.Context) gimlet.Responder {
+	patchMap, err := p.sc.GetPatchRawPatches(p.patchID)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "Database error"))
+	}
+
+	return gimlet.NewTextResponse(patchMap[p.moduleName])
```

Comment:
If patchMap[p.ModuleName] != ok, we should probably return that it's an invalid moduleName

Reply:
Also, should there be a way for users to get the whole map, in case they want to see all modules or they don't necessarily know what modules a patch has?

Reply:
> we should probably return that it's an invalid moduleName

> or they don't necessarily know what modules a patch has?

I wonder if these answer each other: I was envisioning users would query for the modules they're interested in and we'd give them back an empty string for modules which don't have a patch. I could check the project has a matching module, though, and 404 if it doesn't.

Reply:
> should there be a way for users to get the whole map

I wonder if there was a good way to do this. I considered writing back the whole map in json like this:
```
{
"": "main patch diff",
"module1": "module1 diff",
"module2": "module2 diff",
...
}
```
but decided against it because there are cases where the patches can't be encoded in a string, such as patches containing characters that can't be utf-8 encoded (and I didn't want to give users a base64 encoded string), or binary patches.

Reply:
> I could check the project has a matching module, though, and 404 if it doesn't

on second thought, the patch may not have a version/parser project yet. It it worth it to parse the patch's patched_config?

Reply:
> they don't necessarily know what modules a patch has

they could get the list beforehand from the GET /patches/{patch_id} route

Reply:
>  I could check the project has a matching module, though, and 404 if it doesn't.
What I meant was if `val, ok := map["module1"]; !ok` then return an error saying "no patch diff available for this module or something. I don't think the right behavior is to return an empty string.

Reply:
Specifically, this would only happen if the user like misstyped a module or something, or assumed a module exists that doesn't, because we insert into the map for every patch.Patches, so that's why I believe this check is correct.

Reply:
> return an error

If the user asks for a patch_id which doesn't exist, that's an invalid request and we should return a 400 range code. If the user asks for a module that exists in the project we need to return something in the 200 range because there's nothing wrong with the request: it's a _valid question_ whether a patch contains a patch for a given module, even if we happen not to have one.

> you'd likely use StatusBadRequest

A [404](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.5) would be more specific. 400 indicates [malformed syntax](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.1)

Reply:
Resolved in person.